### PR TITLE
Trakt private list support

### PR DESCRIPTION
**Pull request by geogolem**
#### Database Migration
NO

#### Description
This allows users to specify Trakt API Auth Token and Refresh Tokens so that Radarr can access private lists

#### Todos
- [ ] Tests
- [ ] Documentation

#### Issues Fixed or Closed by this PR
http://feathub.com/Radarr/Radarr/+62
* 
This code assumes that the user has already obtained the auth_token and the refresh_token

we still should implement a button somewhere that performs the tasks described here: 

http://docs.trakt.apiary.io/#reference/authentication-oauth

so that radarr can help the user obtain these codes within radarr itself.

### CODE REVIEW

**Comment by geogolem** [comment]
some notes on obtaining the auth token and refresh token:

users should be directed to the following url: https://api.trakt.tv/oauth/authorize?response_type=code&client_id=657bb899dcb81ec8ee838ff09f6e013ff7c740bf0ccfa54dd41e791b9a70b2f0&redirect_uri=urn:ietf:wg:oauth:2.0:oob

if they log in with their trakt account they will receive a PIN..

they should then return to radarr and input this pin and click a button...

then radarr will get the auth key and refresh token automatically using a post request of the following form where code i: 

{
    "code": [pin],
    "client_id": [api-key],
    "client_secret": "[client-secret],
    "redirect_uri": "urn:ietf:wg:oauth:2.0:oob",
    "grant_type": "authorization_code"
}

the returned response should be of the form: 

{
  "access_token": ,
  "token_type": "bearer",
  "expires_in": ,
  "refresh_token": ,
  "scope": "public",
  "created_at": 
}

**Comment by geogolem** [comment]
some notes on obtaining the auth token and refresh token:

http://docs.trakt.apiary.io/#reference/authentication-devices/device-code/generate-new-device-codes
users should click a button, radarr will then issue a POST to https://api.trakt.tv/oauth/device/code

with the BODY: {
  "client_id": [radarr api-key]
}
and will receive a response of the form:

{
  "device_code": [device_code],
  "user_code": [user_code],
  "verification_url": "https://trakt.tv/activate",
  "expires_in": 600,
  "interval": [interval]
}
radarr should then pop-up a dialog box telling the user to visit: [verification_url] and enter [user_code]

this dialog box should have a cancel button (but no OK button)

http://docs.trakt.apiary.io/#reference/authentication-devices/device-code/poll-for-the-access_token

radarr should then use the [device_code] to poll at [interval] in seconds to check if the user has authorized radarr: https://api.trakt.tv/oauth/device/token with a POST of the form:

{
  "code": [device_code],,
  "client_id": [radarr_api_key],
  "client_secret": [client_secret]
}

radarr should handle these stastus codes: Code	Description
200	Success - save the access_token
400	Pending - waiting for the user to authorize your app
404	Not Found - invalid device_code
409	Already Used - user already approved this code
410	Expired - the tokens have expired, restart the process
418	Denied - user explicitly denied this code
429	Slow Down - your app is polling too quickly

and use the returned expires_in field to timeout because the user took too long.

when SUCCESS is received, radarr should populate the auth_token and refresh_token in its configuration and close the pop-up window.

if the user clicks cancelin the pop up window, the polling should end and the process has been canceled.

I could implement this but I dont know how to get a button with a pop-up on the config menu..

It should be similar to "refresh APIKEY".. except instead of a pop up simply warning the user with Cancel, OK... it should have info telling the user what to do, and a cancel button (no ok button) and radarr should be polling until it is successful, the user clicks cancel or the device code expires.

there should also be a revoke button http://docs.trakt.apiary.io/#reference/authentication-oauth/get-token/revoke-an-access_token?console=1

which simply does the POST and then clears the token fields in radarr.

**Comment by geogolem** [comment]
not sure if the client_secret needs to actually be secret.. . but if i cant expose in the code, not sure how to go about handling all this..

**Comment by KaraokeStu** [comment]
Great stuff - thanks for adding this feature - I can then make my profile private again!

File: src/NzbDrone.Core/NetImport/Trakt/TraktRequestGenerator.cs [hunk 1]
```diff
@@ -67,38 +67,49 @@ private IEnumerable<NetImportRequest> GetMovies(string searchParameters)
 
             if (_configService.TraktRefreshToken != null) 
             {
-                TimeSpan span = (DateTime.UtcNow - new DateTime(1970, 1, 1, 0, 0, 0, 0,DateTimeKind.Utc));
-                double unixTime = span.TotalSeconds;
-                             
+                // TimeSpan span = (DateTime.UtcNow - new DateTime(1970, 1, 1, 0, 0, 0, 0,DateTimeKind.Utc));
+                //double unixTime = span.TotalSeconds;
+                Int32 unixTime= (Int32)(DateTime.UtcNow.Subtract(new DateTime(1970, 1, 1))).TotalSeconds;
```
Comment (galli-leo) [inline src/NzbDrone.Core/NetImport/Trakt/TraktRequestGenerator.cs]: There should be an easier method in DateTime to do that.

**Pull request closed without merging**